### PR TITLE
HDDS-4925. Introduce ContainerBalancer in SCM with start/stop capabilities.

### PR DESCRIPTION
**Pull request by siddhantsangwan**
## What changes were proposed in this pull request?

Introducing `ContainerBalancer` with start and stop methods. `ContainerBalancerConfiguration` class contains the necessary config values. Adding `BALANCER` tag in `ConfigTag`. `ContainerBalancer` is constructed when `StorageContainerManager` is created.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-4925

## How was this patch tested?

`TestStorageContainerManagerHA`

### CODE REVIEW

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancerConfiguration.java [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+import org.apache.hadoop.hdds.conf.ConfigGroup;
+import org.apache.hadoop.hdds.conf.ConfigTag;
+import org.apache.hadoop.hdds.conf.ConfigType;
+
+/**
+ * This class contains configuration values for the ContainerBalancer.
+ */
+@ConfigGroup(prefix = "hdds.container.balancer")
+public final class ContainerBalancerConfiguration {
+  @Config(key = "threshold", type = ConfigType.AUTO, defaultValue = "0.1",
```
Comment (lokeshj1703) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancerConfiguration.java:32]: Let's have a prefix "utilisation" or sth more suitable for this key.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancerConfiguration.java [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+      tags = {ConfigTag.BALANCER}, description = "Threshold is a fraction " +
+      "in the range of 0 to 1. A cluster is considered balanced if for each " +
+      "datanode, the utilization of the datanode (used space to capacity " +
+      "ratio) differs from the utilization of the cluster (used " +
+      "space to capacity ratio of the entire cluster) no more than the " +
+      "threshold value.")
+  private double threshold;
+
+  @Config(key = "max.datanodes.balanced", type = ConfigType.INT,
```
Comment (lokeshj1703) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancerConfiguration.java:41]: Lets rename the key to datanodes.balanced.max. max can be suffix. Similarly for the other config.

**Review by lokeshj1703 — COMMENTED** [review]
@siddhantsangwan Thanks for working on this! The changes look good to me. I have few minor comments inline.

**Comment by siddhantsangwan** [comment]
I have incorporated the suggestions and made some other changes to pass the findbugs check. Please let me know if there are any more suggestions.

**Review by lokeshj1703 — APPROVED** [review]
@siddhantsangwan Thanks for updating the PR! The changes look good to me. +1.

**Comment by lokeshj1703** [comment]
@siddhantsangwan Thanks for the contribution! I have committed the PR to master branch.

**Pull request closed without merging**